### DISCUSSION TITLE
Multi-monitor: How can I make sure full screen apps (games) open on the correct monitor?

### DISCUSSION THREAD

**Discussion by winteriscariot** [labwc/labwc#1859] [Q&A]
I have a 3 display setup, and right now it's very unpredictable where full screen games are going to open. This includes using the `-f` flag in gamescope. Right now my best work around for games opening consistently on the correct display is to use gamescope, but even that has its own issues and quirks.

Is there any option or trick I can do to make sure full screen windows open on the right display? My side panels are all rotated differently so a game in a vertical panel won't seem to be able to be moved to a different screen.

Thanks!

**Comment by tokyo4j** [comment]
I tried `gamescope -f` and it opens in the output where the cursor resides. Does it behave differently in your environment?

**Reply by winteriscariot** [reply]
This is inconsistent. Take the follow example:

Main monitor, no windows. Side panel has maximized firefox.

If I right click on the Steam icon in the systray, then select a game to play, once the "launching" dialog disappears focus returns to the previously focused window, in this case Firefox on my side panel. Since the compositor focus is now in the side panel, the game will launch on that monitor instead of what I consider my "primary" (i know wlroots has no concept of primary displays)

I could *maybe* work around this by opening a window on my main panel and focusing that before launching the game, but I admit I have not rigorously tested this. It'd be preferable if labwc could be told which monitor to open all full screen apps on, all the time.

**Comment by johanmalm** [comment]
Would a window-rule with either of these work?

- `<action name="AutoPlace" policy="cursor"/>`
- `<action name="MoveToOutput" output="HDMI-A-1"/>`

https://labwc.github.io/labwc-actions.5.html#entry_action_name=autoplace_policy=value
https://labwc.github.io/labwc-actions.5.html#entry_action_name=movetooutput_output=hdmi-a-1_direction=value_wrap=no

**Reply by winteriscariot** [reply]
Would this work for full screen games? The docs linked for MoveToOutput state: "Moves active window to other output, **unless the window state is fullscreen**."

i'm just trying to make full screen windows (games, in this case) reliably open on the same monitor every time and that just doesn't seem to occur. additionally, games seem to get random resolutions that don't correspond to anything in reality (got a really weird 4,000 pixel wide resolution in Elden Ring)

**Reply by winteriscariot** [reply]
I will try using the AutoPlace cursor rule, I thought I had that set in my rc.xml but it appears I do not.

**Reply by tokyo4j** [reply]
@winteriscariot
I think you're right. We currently don't allow MoveToOutput action for fullscreen window.

If the application opens in windowed mode first and make it fullscreen later, MoveToOutput works. But that causes flickering and many applications would not do that.

We may implement something like `<windowRule event="onCreate">` in the future, by which users can set the position of window before the window is shown. More discussion about this feature will be done in #1718.

**Reply by gzcao** [reply]
@tokyo4j Before onCreate is implemented, what alternatives are there to ensure a fullscreen window opens on a particular output?  Is there a way to move the cursor to a particular output to take advantage of the autoplace cursor policy for the new window?

**Reply by tokyo4j** [reply]
Sorry for late response. Looks like #2040 allowed `MoveToOutput` action to be applied for fullscreen windows. Could you try it again?

**Comment by winteriscariot** [comment]
I've been playing around with this a bit, because the behavior is exceptionally inconsistent. I want to avoid using Gamescope if possible, as there are several issues with that i've run into (and it doesn't work for all games)

Let me describe the current situation, using Path of Exile 2:

* I launch Path of Exile 2 without gamescope. it opens randomly on one of my 4 monitors
* I go into PoE2 settings and the game is smart enough that i can change the Display in Graphics options and the window will move to the correct panel without a fuss
* But if i go to bed and wake up the next morning, without turning my computer off (but having the monitors blank), when i launch the game again it opens, again, on a random monitor

What i notice when i come back to my computer after it goes idle and blanks the monitors, is that the list in PoE2 settings the list of monitors moves around. see the screenshot:

![image](https://github.com/user-attachments/assets/0f8a5d4a-8e39-4030-a235-bd4891931fa4)

currently my primary (centered) 1440p panel is marked as Monitor 3; sometimes, when i launch the game, it'll be marked as Monitor 4, so i'll have to go into settings and change the monitor.

I *think* this is because my 1440p panel initializes from blank slower than the other panels? so the other panels show up first and then this 1440p panel shows whenever it shows up and becomes Monitor 4, when it was 3 previously.

Is there a way to prevent this? is there a way to, idk, "fix" a particular monitor to always show up as, say, DP-1/2/3, HDMI-A-1/2/3 etc?

**Reply by winteriscariot** [reply]
i'm going to try flipping this on and see if it helps:


<core><reuseOutputMode> [yes|no]
Try to re-use the existing output mode (resolution / refresh rate). This may prevent unnecessary screenblank delays when starting labwc (also known as flicker free boot). If the existing output  mode can not be used with labwc the preferred mode of the monitor is used instead. Default is no.

**Reply by Consolatis** [reply]
Setting some xwayland output as primary (with xrandr) before starting the game \*might\* help.
Regarding the wakeup-from-blanked-state issue: some monitors (or display drivers, not sure) have the annoying behavior of disconnecting and reconnecting themselves when being re-enabled. In that case they are completely new monitors from the labwc (and xwayland) perspective so the ordering may in fact differ every time it happens.

**Reply by winteriscariot** [reply]
> have the annoying behavior of disconnecting and reconnecting themselves when being re-enabled. In that case they are completely new monitors from the labwc (and xwayland) perspective so the ordering may in fact differ every time it happens.

would it be possible to get "reliable names" out of monitors, that can associate the raw EDID info with DP-1/HDMI-A1/etc/etc? or would that require whole new code logic to accomplish?

**Reply by Consolatis** [reply]
Internally we do have some things for monitors like `Make`, `Model` and `Serial`. That information is also exposed via the output-management protocol (e.g. `wlr-randr`) but I don't think xwayland uses that protocol. I guess technically one could match the two externally based on the connector name (e.g. `HDMI-A-1`) via parsing of `wlr-randr` and `xrandr`.
